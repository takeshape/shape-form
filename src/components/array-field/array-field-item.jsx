import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import Button from '@material-ui/core/Button';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import ExpansionPanelActions from '@material-ui/core/ExpansionPanelActions';
import ExpandMoreIcon from '@material-ui/icons/ExpandMore';
import Reorder from '@material-ui/icons/Reorder';
import Typography from '@material-ui/core/Typography';
import withStyles from '@material-ui/styles/withStyles';
import SchemaField from '../schema-field';
import sortableItem from '../../decorators/sortable-item';
import ContentPreview from '../content-preview';

class ArrayFieldItem extends PureComponent {
  static propTypes = {
    classes: PropTypes.shape({
      details: PropTypes.string,
      expanded: PropTypes.string,
      dragging: PropTypes.string,
      previewFix: PropTypes.string
    }),
    formName: PropTypes.string.isRequired,
    schema: PropTypes.object.isRequired,
    path: PropTypes.string.isRequired,
    config: PropTypes.object.isRequired,
    context: PropTypes.object,
    widgets: PropTypes.object,
    disabled: PropTypes.bool,
    locale: PropTypes.string,

    // Array related:
    index: PropTypes.number.isRequired,
    arrayKey: PropTypes.number.isRequired,
    isCollapsed: PropTypes.bool.isRequired,
    removeItem: PropTypes.func.isRequired,
    toggleItem: PropTypes.func.isRequired,
    fieldAction: PropTypes.func.isRequired,
    isDragging: PropTypes.bool.isRequired,
    isPlaceholder: PropTypes.bool,
    connectDragSource: PropTypes.func.isRequired,
    connectDropTarget: PropTypes.func.isRequired,
    connectDragPreview: PropTypes.func.isRequired,
    onBlur: PropTypes.func
  };

  handleRemove = event => {
    event.preventDefault();
    const {removeItem, index, onBlur} = this.props;
    removeItem(index);
    onBlur();
  };

  handleToggle = event => {
    event.preventDefault();
    const {toggleItem, arrayKey} = this.props;
    toggleItem(arrayKey);
  };

  setDomNode = node => {
    this.domNode = node;
  };

  render() {
    const {
      classes,
      formName,
      schema,
      path,
      index,
      config,
      context,
      widgets,
      connectDragSource,
      connectDropTarget,
      connectDragPreview,
      isDragging,
      isCollapsed,
      isPlaceholder,
      disabled,
      locale
    } = this.props;

    const currentPath = `${path}[${index}]`;

    let field;
    if (!isCollapsed && !isPlaceholder) {
      field = (
        <SchemaField
          schema={schema}
          formName={formName}
          path={currentPath}
          config={config}
          context={context}
          widgets={widgets}
          disabled={disabled}
          locale={locale}
        />
      );
    }

    const Wrapper = widgets && widgets[config.wrapper];
    if (Wrapper) {
      return (
        <Wrapper
          {...this.props}
          setDomNode={this.setDomNode}
          onRemoveItem={this.handleRemove}
          onToggleItem={this.handleToggle}
          parentPath={path}
          path={currentPath}
          disabled={disabled}
        >
          {field}
        </Wrapper>
      );
    }
    const expansion = (
      <div>
        <ExpansionPanel
          square
          onChange={this.handleToggle}
          expanded={!isCollapsed}
          disabled={disabled}
          classes={{root: isDragging ? classes.dragging : undefined, expanded: classes.expanded}}
        >
          <ExpansionPanelSummary expandIcon={<ExpandMoreIcon />}>
            {!disabled && <Reorder />}
            <Typography>
              <ContentPreview
                formName={formName}
                path={currentPath}
                schema={schema}
                config={config}
                widgets={widgets}
                locale={locale}
              />
            </Typography>
          </ExpansionPanelSummary>
          {field && <ExpansionPanelDetails className={classes.details}>{field}</ExpansionPanelDetails>}
          <ExpansionPanelActions>
            <Button variant="text" onClick={this.handleRemove}>
              Remove
            </Button>
          </ExpansionPanelActions>
        </ExpansionPanel>
      </div>
    );

    return connectDragPreview(
      connectDropTarget(
        <li className={classes.previewFix} ref={this.setDomNode}>
          {disabled ? expansion : connectDragSource(expansion)}
        </li>
      )
    );
  }
}

const styles = {
  details: {
    flexDirection: 'column'
  },
  expanded: {
    marginTop: '0.5rem',
    marginBottom: '0.5rem'
  },
  dragging: {
    visibility: 'hidden'
  },
  previewFix: {
    transform: 'translate3d(0, 0, 0)'
  }
};

export {ArrayFieldItem};
export default withStyles(styles)(sortableItem(ArrayFieldItem));
