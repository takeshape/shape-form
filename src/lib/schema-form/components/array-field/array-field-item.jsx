import PropTypes from 'prop-types';
import React, {PureComponent} from 'react';
import SchemaField from '../schema-field';
import {IconButton} from 'react-toolbox/lib/button';
import {Card, CardText} from 'react-toolbox/lib/card';
import sortableItem from '../../decorators/sortable-item';
import style from './array-field-item.scss';
import classnames from 'classnames';
import ContentPreview from '../content-preview';
import {caretIcon} from './icons';

class ArrayFieldItem extends PureComponent {
  static propTypes = {
    formName: PropTypes.string.isRequired,
    schema: PropTypes.object.isRequired,
    path: PropTypes.string.isRequired,
    config: PropTypes.object.isRequired,
    context: PropTypes.object,
    widgets: PropTypes.object,
    disabled: PropTypes.bool,
    locale: PropTypes.string,

    // Array related:
    index: PropTypes.number.isRequired,
    arrayKey: PropTypes.number.isRequired,
    isCollapsed: PropTypes.bool.isRequired,
    removeItem: PropTypes.func.isRequired,
    toggleItem: PropTypes.func.isRequired,
    fieldAction: PropTypes.func.isRequired,
    isDragging: PropTypes.bool.isRequired,
    isPlaceholder: PropTypes.bool,
    connectDragSource: PropTypes.func.isRequired,
    connectDropTarget: PropTypes.func.isRequired,
    connectDragPreview: PropTypes.func.isRequired,
    onBlur: PropTypes.func
  };

  handleRemove = event => {
    event.preventDefault();
    const {removeItem, index, onBlur} = this.props;
    removeItem(index);
    onBlur();
  };

  handleToggle = event => {
    event.preventDefault();
    const {toggleItem, arrayKey} = this.props;
    toggleItem(arrayKey);
  };

  setDomNode = node => {
    this.domNode = node;
  };

  render() {
    const {
      formName,
      schema,
      path,
      index,
      config,
      context,
      widgets,
      connectDragSource,
      connectDropTarget,
      connectDragPreview,
      isDragging,
      isCollapsed,
      isPlaceholder,
      disabled,
      locale
    } = this.props;

    const currentPath = `${path}[${index}]`;

    let field;
    if (!isCollapsed && !isPlaceholder) {
      field = (
        <SchemaField
          schema={schema}
          formName={formName}
          path={currentPath}
          config={config}
          context={context}
          widgets={widgets}
          disabled={disabled}
          locale={locale}
        />
      );
    }

    const Wrapper = widgets && widgets[config.get('wrapper')];
    if (Wrapper) {
      return (
        <Wrapper
          {...this.props}
          setDomNode={this.setDomNode}
          onRemoveItem={this.handleRemove}
          onToggleItem={this.handleToggle}
          parentPath={path}
          path={currentPath}
          disabled={disabled}
        >
          {field}
        </Wrapper>
      );
    }

    const classNames = classnames(style.element, {
      [style.dragging]: isDragging || isPlaceholder
    });

    const caretButtonStyles = classnames(style.caret, {
      [style.collapsed]: isCollapsed
    });

    const headerClassNames = classnames(style.header, {
      [style.draggable]: !disabled
    });

    const header = (
      <header className={headerClassNames}>
        <IconButton className={caretButtonStyles} onClick={this.handleToggle}>
          {caretIcon}
        </IconButton>
        <ContentPreview
          formName={formName}
          path={currentPath}
          schema={schema}
          config={config}
          widgets={widgets}
          locale={locale}
        />
        <IconButton className={style.delete} icon="delete" onClick={this.handleRemove} disabled={disabled} />
      </header>
    );

    return connectDragPreview(
      connectDropTarget(
        <li className={classNames} ref={this.setDomNode}>
          <Card theme={style}>
            {disabled ? header : connectDragSource(header)}
            {!isCollapsed && <CardText theme={style}>{field}</CardText>}
          </Card>
        </li>
      )
    );
  }
}

export default sortableItem(ArrayFieldItem);
