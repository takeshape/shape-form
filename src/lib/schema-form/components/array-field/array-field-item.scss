@import "~react-toolbox-legacy/lib/colors";

.element {
  opacity: 1;
  list-style: none;

  & + * {
    margin-top: 1rem;
  }
}

.dragging {
  opacity: 0;
}

.card {
  box-shadow: none;
  overflow: visible;
  transition: height 1s;

  & + * {
    margin-top: 4rem;
  }
}

.cardText {
  border-left: 1px solid $palette-blue-grey-50;
  border-bottom: 1px solid $palette-blue-grey-50;
  border-right: 1px solid $palette-blue-grey-50;

  &:last-child {
    padding-bottom: 1.4rem;
  }
}

.header {
  background-color: rgb(238,241,247);
  width: 100%;
  padding: 0.6rem 0.6rem 0.6rem 0;
  color: $palette-blue-grey-500;
  font-size: 1.2rem;

  &.draggable {
    cursor: move;
  }
}

.delete {
  float: right;
}

.caret {
  svg {
    height: 1.2rem;
    margin: 1.2rem 0;
    transition: 200ms ease all;
    transform: rotate(90deg);
  }
}

.collapsed {
  svg {
    transform: rotate(0deg);
  }
}
