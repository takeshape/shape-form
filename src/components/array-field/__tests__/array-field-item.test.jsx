jest.mock('../../schema-field', () => 'SchemaField');
jest.mock('../../content-preview', () => 'ContentPreview');

import React from 'react';
import ReactShallowRenderer from 'react-test-renderer/shallow';
import Button from '@material-ui/core/Button';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import {shallow} from 'enzyme';
import {ArrayFieldItem} from '../array-field-item';

const renderer = new ReactShallowRenderer();

const schema = {
  type: 'object',
  properties: {
    email: {type: 'string'},
    firstName: {type: 'string'},
    lastName: {type: 'string'}
  }
};

function createProps(props) {
  const passthrough = x => x;
  return {
    formName: 'test-form',
    schema,
    path: 'users',
    config: {},
    context: {foo: 'bar'},
    widgets: {},

    // Array related:
    index: 0,
    arrayKey: 0,
    isCollapsed: false,
    removeItem: jest.fn(),
    toggleItem: jest.fn(),
    fieldAction: jest.fn(),
    isDragging: false,
    connectDragSource: passthrough,
    connectDropTarget: passthrough,
    connectDragPreview: passthrough,
    onBlur: jest.fn(),
    locale: 'pt-br',
    classes: {expanded: 'expanded', details: 'details'},
    ...props
  };
}

function mockEvent() {
  return {
    preventDefault: jest.fn()
  };
}

describe('ArrayFieldItem', () => {
  it('renders', () => {
    const props = createProps();
    const tree = renderer.render(<ArrayFieldItem {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders when disabled', () => {
    const props = createProps();
    const tree = renderer.render(<ArrayFieldItem {...props} disabled />);
    expect(tree).toMatchSnapshot();
  });

  it('renders when collapsed', () => {
    const props = createProps({isCollapsed: true});
    props.isCollapsed = true;
    const tree = renderer.render(<ArrayFieldItem {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders with custom wrapper', () => {
    const Unstyled = ({children}) => <div>{children}</div>; // eslint-disable-line
    const props = createProps({
      config: {wrapper: 'unstyled'},
      widgets: {unstyled: Unstyled}
    });
    const tree = renderer.render(<ArrayFieldItem {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('handleToggle', () => {
    const props = createProps();

    const wrapper = shallow(<ArrayFieldItem {...props} />);
    const event = mockEvent();
    wrapper
      .find(ExpansionPanel)
      .at(0)
      .simulate('change', event);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect(props.toggleItem).toHaveBeenCalledTimes(1);
    expect(props.toggleItem).toHaveBeenCalledWith(props.arrayKey);
  });

  it('handleRemove', () => {
    const props = createProps();

    const wrapper = shallow(<ArrayFieldItem {...props} />);
    const event = mockEvent();
    wrapper
      .find(Button)
      .at(0)
      .simulate('click', event);
    expect(event.preventDefault).toHaveBeenCalledTimes(1);
    expect(props.removeItem).toHaveBeenCalledTimes(1);
    expect(props.removeItem).toHaveBeenCalledWith(props.index);
    expect(props.onBlur).toHaveBeenCalledTimes(1);
  });
});
