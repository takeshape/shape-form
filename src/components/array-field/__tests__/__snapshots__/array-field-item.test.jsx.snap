// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArrayFieldItem renders 1`] = `
<li>
  <div>
    <ForwardRef(WithStyles)
      classes={
        Object {
          "expanded": "expanded",
          "root": undefined,
        }
      }
      expanded={true}
      onChange={[Function]}
      square={true}
    >
      <ForwardRef(WithStyles)
        expandIcon={<Memo(ExpandMoreIcon) />}
      >
        <Memo(ReorderIcon) />
        <ForwardRef(WithStyles)>
          <ContentPreview
            config={Object {}}
            formName="test-form"
            locale="pt-br"
            path="users[0]"
            schema={
              Object {
                "properties": Object {
                  "email": Object {
                    "type": "string",
                  },
                  "firstName": Object {
                    "type": "string",
                  },
                  "lastName": Object {
                    "type": "string",
                  },
                },
                "type": "object",
              }
            }
            widgets={Object {}}
          />
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        className="details"
      >
        <SchemaField
          config={Object {}}
          context={
            Object {
              "foo": "bar",
            }
          }
          formName="test-form"
          locale="pt-br"
          path="users[0]"
          schema={
            Object {
              "properties": Object {
                "email": Object {
                  "type": "string",
                },
                "firstName": Object {
                  "type": "string",
                },
                "lastName": Object {
                  "type": "string",
                },
              },
              "type": "object",
            }
          }
          widgets={Object {}}
        />
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)
          onClick={[Function]}
          variant="text"
        >
          Remove
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
  </div>
</li>
`;

exports[`ArrayFieldItem renders when collapsed 1`] = `
<li>
  <div>
    <ForwardRef(WithStyles)
      classes={
        Object {
          "expanded": "expanded",
          "root": undefined,
        }
      }
      expanded={false}
      onChange={[Function]}
      square={true}
    >
      <ForwardRef(WithStyles)
        expandIcon={<Memo(ExpandMoreIcon) />}
      >
        <Memo(ReorderIcon) />
        <ForwardRef(WithStyles)>
          <ContentPreview
            config={Object {}}
            formName="test-form"
            locale="pt-br"
            path="users[0]"
            schema={
              Object {
                "properties": Object {
                  "email": Object {
                    "type": "string",
                  },
                  "firstName": Object {
                    "type": "string",
                  },
                  "lastName": Object {
                    "type": "string",
                  },
                },
                "type": "object",
              }
            }
            widgets={Object {}}
          />
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)
          onClick={[Function]}
          variant="text"
        >
          Remove
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
  </div>
</li>
`;

exports[`ArrayFieldItem renders when disabled 1`] = `
<li>
  <div>
    <ForwardRef(WithStyles)
      classes={
        Object {
          "expanded": "expanded",
          "root": undefined,
        }
      }
      disabled={true}
      expanded={true}
      onChange={[Function]}
      square={true}
    >
      <ForwardRef(WithStyles)
        expandIcon={<Memo(ExpandMoreIcon) />}
      >
        <ForwardRef(WithStyles)>
          <ContentPreview
            config={Object {}}
            formName="test-form"
            locale="pt-br"
            path="users[0]"
            schema={
              Object {
                "properties": Object {
                  "email": Object {
                    "type": "string",
                  },
                  "firstName": Object {
                    "type": "string",
                  },
                  "lastName": Object {
                    "type": "string",
                  },
                },
                "type": "object",
              }
            }
            widgets={Object {}}
          />
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        className="details"
      >
        <SchemaField
          config={Object {}}
          context={
            Object {
              "foo": "bar",
            }
          }
          disabled={true}
          formName="test-form"
          locale="pt-br"
          path="users[0]"
          schema={
            Object {
              "properties": Object {
                "email": Object {
                  "type": "string",
                },
                "firstName": Object {
                  "type": "string",
                },
                "lastName": Object {
                  "type": "string",
                },
              },
              "type": "object",
            }
          }
          widgets={Object {}}
        />
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)>
        <ForwardRef(WithStyles)
          onClick={[Function]}
          variant="text"
        >
          Remove
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
  </div>
</li>
`;

exports[`ArrayFieldItem renders with custom wrapper 1`] = `
<Unstyled
  arrayKey={0}
  classes={
    Object {
      "details": "details",
      "expanded": "expanded",
    }
  }
  config={
    Object {
      "wrapper": "unstyled",
    }
  }
  connectDragPreview={[Function]}
  connectDragSource={[Function]}
  connectDropTarget={[Function]}
  context={
    Object {
      "foo": "bar",
    }
  }
  fieldAction={[MockFunction]}
  formName="test-form"
  index={0}
  isCollapsed={false}
  isDragging={false}
  locale="pt-br"
  onBlur={[MockFunction]}
  onRemoveItem={[Function]}
  onToggleItem={[Function]}
  parentPath="users"
  path="users[0]"
  removeItem={[MockFunction]}
  schema={
    Object {
      "properties": Object {
        "email": Object {
          "type": "string",
        },
        "firstName": Object {
          "type": "string",
        },
        "lastName": Object {
          "type": "string",
        },
      },
      "type": "object",
    }
  }
  setDomNode={[Function]}
  toggleItem={[MockFunction]}
  widgets={
    Object {
      "unstyled": [Function],
    }
  }
>
  <SchemaField
    config={
      Object {
        "wrapper": "unstyled",
      }
    }
    context={
      Object {
        "foo": "bar",
      }
    }
    formName="test-form"
    locale="pt-br"
    path="users[0]"
    schema={
      Object {
        "properties": Object {
          "email": Object {
            "type": "string",
          },
          "firstName": Object {
            "type": "string",
          },
          "lastName": Object {
            "type": "string",
          },
        },
        "type": "object",
      }
    }
    widgets={
      Object {
        "unstyled": [Function],
      }
    }
  />
</Unstyled>
`;
